docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+"stones". He terms the methods which actually build Theano graphs
+application methods (what a coincidence!).
+
+    >>> # D's D2L2v0.4
+    >>> class Stone(object):
+    >>>     """Base class, might have some useful functionality later."""
+    >>>    pass
+    >>>
+    >>> 
+    >>> class MultiLayer(Stone):
+    >>>     def __init__(self, dims, activations, 
+    >>>                  weight_initializers, bias_initializers):
+    >>>         self.dims = dims
+    >>>         self.activations = activations
+    >>>         self.weight_matrices = []
+    >>>         self.bias_vectors = []
+    >>>         for i in range(len(dims) - 1):
+    >>>             self.weight_matrices.append(
+    >>>                 theano.shared(weight_initializers[i]((dims[i], dims[i + 1]))))
+    >>>         for i in range(len(dims) - 1):
+    >>>             self.bias_vectors.append(theano.shared(
+    >>>                 bias_initializers[i](dims[i + 1])))
+    >>>             
+    >>>     def apply(self, inputs):
+    >>>         outputs = inputs
+    >>>         for dim, activation, weight_matrix, bias_vector in zip(
+    >>>                 self.dims[1:], self.activations, self.weight_matrices, self.bias_vectors):
+    >>>             outputs = activation(outputs.dot(weight_matrix) + bias_vector)
+    >>>         return outputs    
+    >>> 
+    >>>     
+    >>> class RecurrentNetwork(Stone):
+    >>>     def __init__(self, dim, weight_initializer):
+    >>>         self.dim = dim
+    >>>         self.weight_initializer = weight_initializer
+    >>>         self.weight_matrix = theano.shared(weight_initializer((dim, dim)))
+    >>>         
+    >>>     def apply_one_step(self, states, inputs, weight_matrix): 
+    >>>         return states.dot(weight_matrix) + inputs
+    >>> 
+    >>>     def apply(self, inputs):
+    >>>         outputs, _ = theano.scan(
+    >>>             self.apply_one_step,
+    >>>             sequences=[inputs], non_sequences=[self.weight_matrix],
+    >>>             outputs_info=[tensor.zeros_like(inputs[0])])
+    >>>         return outputs
+    >>> 
+    >>> 
+    >>> inputs = tensor.matrix('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(10, 754))
+    >>> dims = [754, 100, 100, 10]
+    >>> activations = [tensor.tanh, tensor.tanh, tensor.nnet.softmax]
+    >>> weight_initializers = [
+    >>>     lambda shape: numpy.random.uniform(size=shape)] * len(activations)
+    >>> bias_initializers = weight_initializers
+    >>> stone = MultiLayer(
+    >>>     dims, activations, weight_initializers, bias_initializers)
+    >>> outputs_multilayer = stone.apply(inputs)
+    >>>     
+    >>> inputs = tensor.tensor3('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(5, 10, 100))
+    >>> stone = RecurrentNetwork(100, lambda shape: numpy.random.uniform(size=shape))
+    >>> outputs_recurrent = stone.apply(inputs)
+
+We do not list the hypothetical sampling code here for brevity, but
```

Comment:
Do you mean that I should show that the cost of supporting dropout would be much higher but for annotations? The only way I see to add dropout to `MultiLayer` would be to support it inside its `__init__` and `apply`, and this is ugly and requires quite a lot of code. Although it would be educative to show how ugly it is, I am afraid of amount of extra code required.

Did I understand your proposal?